kv/txn_coord_sender.go

```diff
@@ -729,19 +732,10 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
 	case *roachpb.OpRequiresTxnError:
 		panic("OpRequiresTxnError must not happen at this level")
 	case *roachpb.ReadWithinUncertaintyIntervalError:
-		// Mark the host as certain. See the protobuf comment for
-		// Transaction.CertainNodes for details.
-		if t.NodeID == 0 {
-			panic("no replica set in header on uncertainty restart")
-		}
-		// No more restarts for this node for anything after ExistingTimestamp.
-		newTxn.UpdateObservedTimestamp(t.NodeID, t.ExistingTimestamp)
 		// If the reader encountered a newer write within the uncertainty
-		// interval, move the timestamp forward, just past that write or
-		// up to ObservedTimestamp, whichever comes first.
-		candidateTS := newTxn.ObservedTimestamp
-		candidateTS.Backward(t.ExistingTimestamp.Add(0, 1))
-		newTxn.Timestamp.Forward(candidateTS)
+		// interval, we simply move past the last observed timestamp from the
+		// node.
+		newTxn.Timestamp.Forward(newTxn.GetObservedTimestamp(pErr.OriginNode).Add(0, 1))
```

Comment:
Actually don't need .Next() at all, removed. Always a bit sketchy to
synthesize timestamps.

On Thu, Mar 3, 2016 at 10:39 AM Spencer Kimball <notifications@github.com>
wrote:

> In kv/txn_coord_sender.go
> <https://github.com/cockroachdb/cockroach/pull/4814#discussion_r54897069>:
>
> >  		// If the reader encountered a newer write within the uncertainty
> > -		// interval, move the timestamp forward, just past that write or
> > -		// up to ObservedTimestamp, whichever comes first.
> > -		candidateTS := newTxn.ObservedTimestamp
> > -		candidateTS.Backward(t.ExistingTimestamp.Add(0, 1))
> > -		newTxn.Timestamp.Forward(candidateTS)
> > +		// interval, we simply move past the last observed timestamp from the
> > +		// node.
> > +		newTxn.Timestamp.Forward(newTxn.GetObservedTimestamp(pErr.OriginNode).Add(0, 1))
>
> Use Timestamp.Next() here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4814/files#r54897069>.
>